### ISSUE TITLE
Summing over discrete latent variables II

### ISSUE THREAD

**Issue by fritzo** [pyro-ppl/pyro#220]
This issue follows up #99 with specific bugfixes and conceptual issues that are needed to get the proposed solution to work.

## Outline

There are three versions of summing-out that we can implement, with increasing use of GPU:
1. `batch_size=1` (#206): DONE 
2. `batch_size>1`: (#256): DONE
3. `BranchPoutine` (#227): needs full broadcasting support in `pyro.distributions` (#231) 

We'd like to support (1,2) by launch, and be on a clear path to implementing (3).

## Tasks

- [x] #223 Fix a bug in non-vectorized `Categorical.support()`, which yields values of the wrong shape.
- [x] #223 When vectorized, `Categorical.support()` iterates over the cartesian product. We'll need to implement a custom version that iterates in lock-step when inside a `map_data()`. Maybe the right thing to do is to implement a special wrapper around discrete distributions, rather than relying on the `.enumerable` field and `.support()` methods.
- [x] #253 Fix sample-shape bugs in `Cagetorical.sample()` and `.batch_log_pdf()`.
- [x] #286, #298 Fix `Categorical.batch_log_pdf()` shape bugs (#278)
- [x] #206 Implement `batch_size=1` summing in `pyro.infer.SVI`.
- [x] #256 Implement summing for `batch_size>1` in `pyro.infer.SVI`.
  Note that trace `scale`s only make sense in the non-vectorized case; in case of vectorized `p` parameters of `Categorical`, each item in the batch should get a different weight. To get this working, we'll need to allow for vectorized `scale` (generalizing #204), and force a final `loss = loss.sum()` after aggregating over particles in `loss_and_grad()`.
- TRIAGED #227 Implement single-trace summing via `BranchPoutine`.
